### PR TITLE
BUG: set index of DataFrame.apply(f) when f returns dict (#37544)

### LINKED ISSUE #37544: BUG: Agg and Apply with custom function improperly ignore column labels when function returns a dict-like
Hello folks! I've discovered an odd corner case regression that I hope can be fixed.

The problem is that if you are running an aggregation/apply on a dataframe using a custom function - if that function returns a dict-like object, it doesn't use the appropriate column labels in the original dataframe.

As a simple replication 
```{python} 
df = pd.DataFrame([[1, 2, 3],
                   [4, 5, 6],
                   [7, 8, 9],
                   [np.nan, np.nan, np.nan]],
                  columns=['A', 'B', 'C'])

print(df.agg(lambda x : {"sum" :np.sum(x)}))
```

prints 
```
0    {'sum': 12.0}
1    {'sum': 15.0}
2    {'sum': 18.0}
```

instead of the expected
```
A    {'sum': 12.0}
B    {'sum': 15.0}
C    {'sum': 18.0}
```

This is a regression/breaking change since in prior versions of pandas (I believe 1.0.X and earlier) the expected results were produced.

### PR DESCRIPTION
**Pull request by ma3da**
- [x] closes #37544
- [x] tests (added) updated / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm can you add a whatsnew note in bug fixes for 1.2 in the reshape section.

**Comment by ma3da** [comment]
Thanks, sure.

File: doc/source/whatsnew/v1.2.0.rst [hunk 1]
```diff
@@ -536,6 +536,7 @@ Reshaping
 - Bug in :func:`join` returned a non deterministic level-order for the resulting :class:`MultiIndex` (:issue:`36910`)
 - Bug in :meth:`DataFrame.combine_first()` caused wrong alignment with dtype ``string`` and one level of ``MultiIndex`` containing only ``NA`` (:issue:`37591`)
 - Fixed regression in :func:`merge` on merging DatetimeIndex with empty DataFrame (:issue:`36895`)
+- Bug in :meth:`FrameRowApply.wrap_results_for_axis` not setting index of return value when ``results`` argument has ``dict`` values. Affects :meth:`DataFrame.apply`. (:issue:`37544`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.2.0.rst:539]: pls make a user facing whatsnew note.

Comment (jreback) [reply doc/source/whatsnew/v1.2.0.rst:539]: you are mentioning internal methods, just say apply had this bug.....

**Review by jreback — CHANGES_REQUESTED** [review]
merge master as well

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @ma3da

**Pull request merged**